src/envs/cluttered_table.py

```diff
@@ -0,0 +1,188 @@
+        # Types
+        self._can_type = Type(
+            "can", ["pose_x", "pose_y", "radius", "is_grasped", "is_trashed"])
+        # Predicates
+        self._HandEmpty = Predicate(
+            "HandEmpty", [], self._HandEmpty_holds)
+        self._Holding = Predicate(
+            "Holding", [self._can_type], self._Holding_holds)
+        # Options
+        self._Grasp = ParameterizedOption(
+            "Grasp", [self._can_type], params_space=Box(0, 1, (4,)),
+            _policy=lambda s, o, p: Action(p),  # action is just parameter
+            _initiable=lambda s, o, p: True,  # can be run from anywhere
+            _terminal=lambda s, o, p: True)  # always 1 timestep
+        self._Dump = ParameterizedOption(
+            "Dump", [], params_space=Box(0, 1, (0,)),
+            _policy=lambda s, o, p: Action(
+                np.zeros(4, dtype=np.float32)),
+            _initiable=lambda s, o, p: True,  # can be run from anywhere
+            _terminal=lambda s, o, p: True)  # always 1 timestep
+        # Objects
+        self._cans = []
+        for i in range(max(CFG.cluttered_table_num_cans_train,
+                           CFG.cluttered_table_num_cans_test)):
+            self._cans.append(Object(f"can{i}", self._can_type))
+
+    def simulate(self, state: State, action: Action) -> State:
+        assert self.action_space.contains(action.arr)
+        next_state = state.copy()
+        # Figure out which can is currently grasped, if any.
+        grasped_can = None
+        for can in state:
+            if state.get(can, "is_grasped") > 0.5:
+                assert grasped_can is None, "Multiple cans grasped?"
+                assert state.get(can, "is_trashed") < 0.5, \
+                    "Grasped a can that has been trashed?"
+                grasped_can = can
+        if np.all(action.arr == 0.0):
+            # Handle dumping action.
+            if grasped_can is not None:
+                next_state.set(grasped_can, "pose_x", -999)
+                next_state.set(grasped_can, "pose_y", -999)
+                next_state.set(grasped_can, "is_grasped", 0.0)
+                next_state.set(grasped_can, "is_trashed", 1.0)
+            return next_state
+        # Handle grasping action.
+        if grasped_can is not None:
+            return next_state  # can't grasp while already grasping
+        start_x, start_y, end_x, end_y = action.arr
+        desired_can = None
+        for can in state:
+            this_x = state.get(can, "pose_x")
+            this_y = state.get(can, "pose_y")
+            this_radius = state.get(can, "radius")
+            if np.linalg.norm([end_x-this_x,
+                               end_y-this_y]) < this_radius:  # type: ignore
+                assert desired_can is None
+                desired_can = can
+        if desired_can is None:
+            return next_state  # end point wasn't at any can
+        vec1 = np.array([end_x-start_x, end_y-start_y])
+        # Handle collision checking.
```

Comment:
can you write some more comments for this logic? in particular I'm surprised that we don't need to use the other can's radius